pinot-common/src/main/java/org/apache/pinot/common/utils/ArrayListUtils.java

```diff
@@ -0,0 +1,124 @@
+    long[] longArrayListElements = longArrayList.elements();
+    return longArrayListElements.length == longArrayList.size() ? longArrayListElements : longArrayList.toLongArray();
+  }
+
+  /**
+   * Best effort extract the given {@link FloatArrayList} to a float array without copying the elements.
+   * The {@link FloatArrayList#elements()} returned float array may be longer than the actual size of the
+   * {@link FloatArrayList}, and the actual size of the {@link FloatArrayList} can be retrieved using
+   * {@link FloatArrayList#size()}.
+   * This method checks the length of the returned float array and returns the same if it is equal to the size of the
+   * {@link FloatArrayList}, otherwise, it copies the elements to a new float array and returns it.
+   *
+   * <p>Use this method only if you are sure that the returned float array will not be modified.
+   * <p>Otherwise, use {@link FloatArrayList#toFloatArray()}.
+   *
+   * @param floatArrayList Input {@link FloatArrayList}
+   * @return Best effort extracted float array without copying the elements
+   */
+  public static float[] toFloatArray(FloatArrayList floatArrayList) {
+    float[] floatArrayListElements = floatArrayList.elements();
+    return floatArrayListElements.length == floatArrayList.size() ? floatArrayListElements
+        : floatArrayList.toFloatArray();
+  }
+
+  /**
+   * Best effort extract the given {@link DoubleArrayList} to a double array without copying the elements.
+   * The {@link DoubleArrayList#elements()} returned double array may be longer than the actual size of the
+   * {@link DoubleArrayList}, and the actual size of the {@link DoubleArrayList} can be retrieved using
+   * {@link DoubleArrayList#size()}.
+   * This method checks the length of the returned double array and returns the same if it is equal to the size of the
+   * {@link DoubleArrayList}, otherwise, it copies the elements to a new double array and returns it.
+   *
+   * <p>Use this method only if you are sure that the returned double array will not be modified.
+   * <p>Otherwise, use {@link DoubleArrayList#toDoubleArray()}.
+   *
+   * @param doubleArrayList Input {@link DoubleArrayList}
+   * @return Best effort extracted double array without copying the elements
+   */
+  public static double[] toDoubleArray(DoubleArrayList doubleArrayList) {
+    double[] doubleArrayListElements = doubleArrayList.elements();
+    return doubleArrayListElements.length == doubleArrayList.size() ? doubleArrayListElements
+        : doubleArrayList.toDoubleArray();
+  }
+
+  /**
+   * Convert the given {@link ObjectArrayList} to a string array.
+   * This method always copies the elements to a new string array and returns it.
+   *
+   * @param stringArrayList Input {@link ObjectArrayList}
+   * @return Copied string array
+   */
+  public static String[] toStringArray(ObjectArrayList<String> stringArrayList) {
+    return stringArrayList.toArray(new String[0]);
```

Comment:
Why not use the same logic as other types to avoid copy?

Reply:
`ObjectArrayList#elements()` might return either Object[] or String[] depends on the interal implementation.
So the type cast might not work always.

Updated this logic to check this scenario.